### ISSUE TITLE
Canary unable to recover from i/o timeout error when trying to describe cluster, keeps trying and failing indefinitely

### ISSUE THREAD

**Issue by grdryn** [strimzi/strimzi-canary#162]
In the same environment ass #159 (using the same canary version 0.2.0), we've hit a different case, where the canary appears to get stuck in an i/o timeout loop:

```
E0121 13:14:45.332810       1 consumer.go:92] Error received whilst consuming from topic: kafka: error while consuming __redhat_strimzi_canary/0: read tcp 10.130.96.13:44774->34.237.155.239:443: i/o timeout
E0121 13:14:46.956177       1 topic.go:130] Error describing cluster: read tcp 10.130.96.13:45720->34.237.155.239:443: i/o timeout
I0121 13:14:46.956217       1 canary_manager.go:134] ... reconcile done
I0121 13:14:46.956224       1 canary_manager.go:124] Canary manager reconcile ...
E0121 13:14:46.956342       1 topic.go:130] Error describing cluster: read tcp 10.130.96.13:45720->34.237.155.239:443: i/o timeout
I0121 13:14:46.956356       1 canary_manager.go:134] ... reconcile done
I0121 13:14:51.955601       1 canary_manager.go:124] Canary manager reconcile ...
E0121 13:14:51.955719       1 topic.go:130] Error describing cluster: read tcp 10.130.96.13:45720->34.237.155.239:443: i/o timeout
I0121 13:14:51.955744       1 canary_manager.go:134] ... reconcile done
I0121 13:14:56.956169       1 canary_manager.go:124] Canary manager reconcile ...
E0121 13:14:56.956302       1 topic.go:130] Error describing cluster: read tcp 10.130.96.13:45720->34.237.155.239:443: i/o timeout
I0121 13:14:56.956320       1 canary_manager.go:134] ... reconcile done
I0121 13:15:01.956278       1 canary_manager.go:124] Canary manager reconcile ...
E0121 13:15:01.956404       1 topic.go:130] Error describing cluster: read tcp 10.130.96.13:45720->34.237.155.239:443: i/o timeout
I0121 13:15:01.956425       1 canary_manager.go:134] ... reconcile done
I0121 13:15:06.956233       1 canary_manager.go:124] Canary manager reconcile ...
E0121 13:15:06.956358       1 topic.go:130] Error describing cluster: read tcp 10.130.96.13:45720->34.237.155.239:443: i/o timeout
I0121 13:15:06.956375       1 canary_manager.go:134] ... reconcile done
I0121 13:15:11.956237       1 canary_manager.go:124] Canary manager reconcile ...
E0121 13:15:11.956371       1 topic.go:130] Error describing cluster: read tcp 10.130.96.13:45720->34.237.155.239:443: i/o timeout
I0121 13:15:11.956388       1 canary_manager.go:134] ... reconcile done
I0121 13:15:16.955833       1 canary_manager.go:124] Canary manager reconcile ...
E0121 13:15:16.955960       1 topic.go:130] Error describing cluster: read tcp 10.130.96.13:45720->34.237.155.239:443: i/o timeout
I0121 13:15:16.955981       1 canary_manager.go:134] ... reconcile done
I0121 13:15:21.956100       1 canary_manager.go:124] Canary manager reconcile ...
E0121 13:15:21.956217       1 topic.go:130] Error describing cluster: read tcp 10.130.96.13:45720->34.237.155.239:443: i/o timeout
I0121 13:15:21.956235       1 canary_manager.go:134] ... reconcile done
E0121 13:15:26.574661       1 consumer.go:92] Error received whilst consuming from topic: kafka: error while consuming __redhat_strimzi_canary/0: write tcp 10.130.96.13:53876->34.237.155.239:443: i/o timeout
I0121 13:15:26.955957       1 canary_manager.go:124] Canary manager reconcile ...
E0121 13:15:26.956068       1 topic.go:130] Error describing cluster: read tcp 10.130.96.13:45720->34.237.155.239:443: i/o timeout
I0121 13:15:26.956090       1 canary_manager.go:134] ... reconcile done
I0121 13:15:31.956163       1 canary_manager.go:124] Canary manager reconcile ...
E0121 13:15:31.956285       1 topic.go:130] Error describing cluster: read tcp 10.130.96.13:45720->34.237.155.239:443: i/o timeout
I0121 13:15:31.956300       1 canary_manager.go:134] ... reconcile done
I0121 13:15:36.956169       1 canary_manager.go:124] Canary manager reconcile ...
E0121 13:15:36.956291       1 topic.go:130] Error describing cluster: read tcp 10.130.96.13:45720->34.237.155.239:443: i/o timeout
I0121 13:15:36.956308       1 canary_manager.go:134] ... reconcile done
[...]
```

This continued for about an hour until it was noticed. Restarting the canary appears to have resolved the issue, so it seems like something the canary should be able to handle itself.

**Comment by k-wall** [comment]
I have a simple way to reproduce a problem with the same symptoms.  Running outside the container (single kafka broker, canary running in the IDE), if you send a STOP signal to your Java Kafka process, I get:

```
E0124 13:15:13.893867    7342 topic.go:130] Error describing cluster: read tcp [::1]:51925->[::1]:9092: i/o timeout
I0124 13:15:13.893949    7342 canary_manager.go:134] ... reconcile done
I0124 13:15:13.893974    7342 canary_manager.go:124] Canary manager reconcile ...
E0124 13:15:13.894027    7342 topic.go:130] Error describing cluster: read tcp [::1]:51925->[::1]:9092: i/o timeout
I0124 13:15:13.894034    7342 canary_manager.go:134] ... reconcile done
```

**Comment by k-wall** [comment]
i/o timeouts (`ErrDeadlineExceeded`) are not necessarily fatal errors.   The sarama code extends the read/write deadline, so it is possible that the next I/O operator will be successful.

https://github.com/Shopify/sarama/blob/f1ea13a383925df424cf52cbb0be0db92032864e/broker.go#L864

I think killing the connection on an ErrDeadlineExceeded condition (at least immediately) is probably the wrong thing to do.

**Comment by ppatierno** [comment]
Closed via #165
